### DISCUSSION TITLE
How to run demo 4

### DISCUSSION THREAD

**Discussion by loneblues** [Aperivue/medsci-skills#433] [Q&A]
Hello. 

Other than the order to use the skills, any specific directions on how to prompt or give input to replicate the demo would be greatly appreciated! 

Also how to apply to a real world that has to be processed, and needs many things to determine, will be very helpful.

Last, if there is a built in labeler which can exploit various interactive models such as you mentioned for labeling to complete an active learning pipeline would be interesting, if it is within the scope of this repo. 

Thanks again for the amazing repo. 

Best, 
HJ Lee

**Comment by Yoojin-nam** [comment]
Hello, and sorry for the slow reply — thank you for the kind words and for three questions that are all sharper than they look.

## 1. The prompts that produced Demo 4

The demo was not run with one prompt. It is the **model-engineering lane**, and the stage-by-stage record of what was invoked is checked into the repository: [`demo/04_pneumoniamnist_cnn/qc/_pipeline_log.md`](https://github.com/Aperivue/medsci-skills/blob/main/demo/04_pneumoniamnist_cnn/qc/_pipeline_log.md). That table is the answer to "what order, and what came out of each step" — it names the skill, the artifact it wrote, and the verdict.

To re-run the computation without any prompting at all:

```bash
cd demo/04_pneumoniamnist_cnn
python3 -m venv .venv && . .venv/bin/activate     # Python 3.11-3.13
pip install -r pipeline/requirements.txt matplotlib captum
bash reproduce.sh
```

To drive it as a conversation on your own task instead, the sequence is roughly this — one prompt per stage, each one naming the skill:

1. `/architecture-zoo` — "binary classification, 2-D chest radiographs, ~5k images, class imbalance about 3:1. What should I start from and why?"
2. `/model-scaffold` — "classification task, the architecture from the previous step, patient-level split, emit the split assignment table."
3. `/model-validation` — "run the split-leakage gate against my split table, strict."
4. `/model-scaffold` — "run the training-hygiene gate, strict." (These two gates are the point of the lane. They are deterministic scripts, not opinions, and they are what makes the later numbers worth anything.)
5. Train. The scaffold emits runnable code; the demo used three seeds so the reported figure carries a seed SD rather than one lucky run.
6. `/model-evaluation` — "held-out metrics for a classification task at deployment prevalence, bootstrap CIs."
7. `/explainability` — "Grad-CAM with both Adebayo sanity checks, quantified."
8. `/write-paper`, then `/self-review`, then `/verify-refs`.

Two things are worth copying more than the prompts. **Every number in that demo came out of an executed run** — none was typed in by the model. And the calibration result (ECE 0.127 next to an AUROC of 0.969) was left in the write-up as a stated weakness rather than quietly dropped. A pipeline that cannot embarrass you is not doing its job.

## 2. Applying it to real data

Demo 4 is the easy case on purpose: a public benchmark with an official split and no patients. For real data the shape changes in four places.

- **Before any modelling**, `/profile-imaging` reads the dataset and tells you what is actually in it — spacing spread, intensity domain, which label values are present, how small the target is — and then audits your stated plan against it. Most of the decisions you are describing as "many things to determine" are answered here rather than guessed.
- **The split must be patient-level**, not image-level. `/model-validation`'s leakage gate proves disjointness by set arithmetic on the emitted split table; it is the single highest-value gate in the toolkit for clinical data.
- **`/preprocess-imaging`** catches the leak a split table cannot see: any transform fitted on data outside the training split. [Demo 5](https://github.com/Aperivue/medsci-skills/tree/main/demo/05_msd_amos_spleen) is built around exactly this — it shows the obvious way of using nnU-Net *failing* that gate, and `reproduce.sh` asserts the failure. If you read one demo for real-world transfer, read that one rather than Demo 4.
- **If the data leaves your machine**, `/deidentify` runs locally with no network and no model call on the raw data.

`/design-study` is the right entry point when the question is still open, and `/orchestrate` will route a request you cannot map to a skill yourself.

## 3. A built-in labeler / active-learning loop

Honest answer: **no, and not today.** The toolkit today *evaluates* interactive and promptable segmentation — `/model-evaluation` carries the axes a static Dice omits (interaction count, convergence, per-case time), and `/architecture-zoo` covers SAM / MedSAM / nnInteractive / VISTA3D as architectures — but nothing in the repository runs an annotation loop or an acquisition function. There is no labeling UI and no active-learning scheduler.

Whether it *should* be in scope is a fair question and I do not think the answer is obviously no. The line the repository has held so far is that skills **design, audit and gate** work, and delegate execution to the established tool (MONAI, nnU-Net, captum, scikit-learn) rather than reimplementing it. An active-learning skill that fitted that line would look like: a gate on the *protocol* — who labelled what, was the acquisition function fitted on the evaluation set, is the label-quality and inter-rater story auditable, does the growing training set stay patient-disjoint from the test set — and would hand the actual labelling to an existing annotator. That framing I would take a PR for.

Would you open an issue describing the loop you have in mind? Concretely: what you are labelling, which model you would put in the loop, and what you would need the toolkit to *check* rather than *do*. That is the part I would most like to get right before anyone writes code.
